### PR TITLE
Refactor and redesign confirm transaction views

### PR DESCRIPTION
**Pull request by alextsg**

### CODE REVIEW

**Comment by danjm** [comment]
@alextsg What are your thoughts on adding directory specific `.selectors` files for most of the `components/pages/confirm-**` directories, to be used to replace any selection logic within the `mapStateToProps` functions?

My preference is for no selection logic at all in `mapStateToProps`, and instead contain all of that in imported selector functions. Also, if there are selectors that will only be used in one container component, my preference is that they be in a file in the same directory as that component.

Thoughts?

File: ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+import { DEFAULT_ROUTE } from '../../../routes'
+import { conversionGreaterThan } from '../../../conversion-util'
+import { MIN_GAS_LIMIT_DEC } from '../../send_/send.constants'
+
+export default class ConfirmTransactionBase extends Component {
+  static contextTypes = {
+    t: PropTypes.func,
+  }
+
+  static propTypes = {
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: It might be good to distinguish, at least with a comment, which of these we expect to get from the container's connection to redux, and which we expect to get passed from the parent.

Comment (danjm) [inline ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: Perhaps alphabetize this list and put spaces between the sections (by type of source) of props.

File: ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+      balance,
+      conversionRate,
+      txData: {
+        txParams: {
+          value: amount,
+        } = {},
+      } = {},
+    } = this.props
+
+    const INSUFFICIENT_FUNDS_ERROR = t('insufficientFunds')
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: I think it might be better to store errors as their translation keys, as opposed to their translations. And then translate them at the point of render.

Reasons:
(1) Data stored will not be a 'view' value, just data that represents the current state of errors
(2) Debugging will be easier if the users state logs are all in english, as opposed to some parts of state that representing error handling being in a different laguage

Also, perhaps these errors should be moved to a constants file

File: ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+        onHide: () => {
+          clearConfirmTransaction()
+          history.push(DEFAULT_ROUTE)
+        },
+      })
+
+      return
+    }
+  }
+
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: What are your thoughts on moving some or all of the logic in `getError` and `validateEditGas` to either reducers or utils?

File: ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+  handleEditGas () {
+    const { onEditGas, showCustomizeGasModal, txData, updateGasAndCalculate } = this.props
+
+    if (onEditGas) {
+      onEditGas()
+    } else {
+      showCustomizeGasModal({
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: Maybe put all of `showCustomizeGasModal` in the container?

File: app/scripts/controllers/transactions/tx-state-manager.js [hunk 5]
```diff
@@ -399,19 +399,17 @@ class TransactionStateManager extends EventEmitter {
   _setTxStatus (txId, status) {
     const txMeta = this.getTx(txId)
     txMeta.status = status
```
Comment (danjm) [inline app/scripts/controllers/transactions/tx-state-manager.js]: Why is this change needed?

File: ui/app/actions.js [hunk 6]
```diff
@@ -705,11 +705,10 @@ function signTypedMsg (msgData) {
 
 function signTx (txData) {
   return (dispatch) => {
-    dispatch(actions.showLoadingIndication())
```
Comment (danjm) [inline ui/app/actions.js]: Why do you remove `showLoadingIndication` and `hideLoadingIndication` here, but not in the other methods edited below?

File: ui/app/actions.js [hunk 7]
```diff
@@ -904,29 +903,41 @@ function signTokenTx (tokenAddress, toAddress, amount, txData) {
...
+          dispatch(actions.txError(err))
+          dispatch(actions.goHome())
+          log.error(err.message)
+          return reject(err)
+        }
+
+        resolve(txData)
+      })
     })
+    .then(() => updateMetamaskStateFromBackground())
```
Comment (danjm) [inline ui/app/actions.js]: Not related to this PR, but we should probably rename `updateMetamaskStateFromBackground` to something like `getUpdatedMetamaskStateFromBackground`. The current name makes it sound like the method does the updating of react state itself.

File: ui/app/selectors/confirm-transaction.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import { createSelector } from 'reselect'
```
Comment (danjm) [inline ui/app/selectors/confirm-transaction.js]: Do you prefer to have these here in a general `selectors/` directory, as opposed to in one of the `confirm-transaction` directories? (As is done, for example, with send)

File: ui/app/actions.js (hunk 7 above)
Comment (danjm) [reply ui/app/actions.js]: Leaving a TODO here for myself to do this after I finish reviwing:

- [ ] create github issue to rename updateMetamaskStateFromBackground

File: ui/app/components/tabs/tabs.component.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    if (tabIndex !== activeTabIndex) {
+      this.setState({
+        activeTabIndex: tabIndex,
+      })
+    }
+  }
+
+  renderTabs () {
+    // const { children } = this.props
+    const numberOfTabs = React.Children.count(this.props.children)
```
Comment (danjm) [inline ui/app/components/tabs/tabs.component.js]: Out of curiosity why is `React.Children` (and not `this.props.children`) needed here?

Comment (danjm) [reply ui/app/components/tabs/tabs.component.js]: - [ ] remove commented out code

File: ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    } else {
+      return detailsComponent || dataComponent
+    }
+  }
+
+  renderTabs () {
+    const { detailsComponent, dataComponent } = this.props
+
+    return (
+      <Tabs>
```
Comment (danjm) [inline ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js]: It might be nice if the tabs component had an API that allowed:
```
<Tabs
  tabs={
    'Details': detailsComponent,
    'Data': dataComponent,
  }
/>
```

File: ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+              hideSubtitle={hideSubtitle}
+              identiconAddress={identiconAddress}
+              nonce={nonce}
+            />
+          )
+        }
+        { this.renderContent() }
+        {
+          errorMessage && (
+            <div className="confirm-page-container-content__error-container">
```
Comment (danjm) [inline ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js]: Should we put the `<div className="confirm-page-container-content__error-container">` inside the `ConfirmPageContainerError` component?

File: ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-summary/confirm-page-container-summary.component.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+              address={identiconAddress}
+            />
+          )
+        }
+        <div className="confirm-page-container-summary__title-text">
+          { title }
+        </div>
+      </div>
+      {
+        hideSubtitle || <div className="confirm-page-container-summary__subtitle">
```
Comment (danjm) [inline ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-summary/confirm-page-container-summary.component.js]: It might be slight improvement on code readability if this was `showSubtitle && `. I say this because the other conditionally rendered elements are rendered if a value is truthy with `&&`. Reading the file and then encountering `||` is a little confusing. I say this lightly though, if switching to `showSubtitle &&` makes things less readable or intuitive elsewhere, then my suggestion here should be ignored.

File: ui/app/components/modals/modal.js [hunk 13]
```diff
@@ -267,7 +269,31 @@ const MODALS = {
 
   CUSTOMIZE_GAS: {
     contents: [
```
Comment (danjm) [inline ui/app/components/modals/modal.js]: I guess we also need to switch over the modal used on the send screen at some point (but not as part of this PR).

- [ ] Create github issue for using new customize gas modal component on send screen.

File: ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js (hunk 4 above)
Comment (danjm) [reply ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: ... maybe not

File: ui/app/components/pages/confirm-transaction-switch/confirm-transaction-switch.container.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import { connect } from 'react-redux'
+import ConfirmTransactionSwitch from './confirm-transaction-switch.component'
+
+const mapStateToProps = state => {
+  const { confirmTransaction } = state
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction-switch/confirm-transaction-switch.container.js]: Perhaps we should map the specific properties within `confirmTransaction` to props here. It will simplify the logic within the component.

File: ui/app/components/pages/confirm-transaction-switch/confirm-transaction-switch.util.js [hunk 15]
```diff
@@ -0,0 +1,4 @@
+export function isConfirmDeployContract (txData = {}) {
+  const { txParams = {} } = txData
+  return !txParams.to
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction-switch/confirm-transaction-switch.util.js]: This is consistent with current code, but I wonder if there is a more specific check. It seems likely that we will be handling other transactions in the future that don't have a recipient address. Perhaps we can parse something from `data`?

File: ui/app/components/pages/confirm-transaction/confirm-transaction.component.js [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    const {
+      history,
+      unconfirmedTransactions,
+      match: { params: { id: paramsTransactionId } = {} },
+      setTransactionToConfirm,
+    } = this.props
+
+    if (paramsTransactionId) {
+      // Check to make sure params ID is valid
+      const tx = R.find(({ id }) => id + '' === paramsTransactionId)(unconfirmedTransactions)
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction/confirm-transaction.component.js]: I'm fine with using ramda, but the code might become a little more accessible if we use the native `find` method.

File: ui/app/components/pages/confirm-transaction/confirm-transaction.component.js [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    if (paramsTransactionId && transactionId && paramsTransactionId !== transactionId + '') {
+      setTransactionToConfirm(paramsTransactionId)
+    }
+
+    if (!transactionId) {
+      this.setTransactionToConfirm()
+    }
+  }
+
+  setTransactionToConfirm () {
```
Comment (danjm) [inline ui/app/components/pages/confirm-transaction/confirm-transaction.component.js]: I think this method is another case where the logic could be moved to the `.container` component and then to selectors.

File: ui/app/components/send_/send.utils.js [hunk 18]
```diff
@@ -45,9 +45,9 @@ function calcGasTotal (gasLimit, gasPrice) {
 
 function isBalanceSufficient ({
   amount = '0x0',
-  amountConversionRate = 0,
```
Comment (danjm) [inline ui/app/components/send_/send.utils.js]: I wonder why `amountConversionRate` was defaulted to `0` before? We should confirm that changing the default to `1` does not break any desired behaviour.

File: ui/app/conversion-util.js [hunk 19]
```diff
@@ -28,6 +28,8 @@ const BN = ethUtil.BN
 const R = require('ramda')
 const { stripHexPrefix } = require('ethereumjs-util')
 
+global.BigNumber = BigNumber
```
Comment (danjm) [inline ui/app/conversion-util.js]: Why add `BigNumber` to global?

File: ui/app/helpers/confirm-transaction/util.js [hunk 20]
```diff
@@ -0,0 +1,116 @@
+import currencyFormatter from 'currency-formatter'
```
Comment (danjm) [inline ui/app/helpers/confirm-transaction/util.js]: I asked a similar question elsewhere. Your thoughts on including this in a general ` ui/app/helpers/` directory as opposed to within a root directory for confirm-transaction components?

File: ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js (hunk 3 above)
Comment (alextsg) [reply ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: I think keeping errors in the redux state makes sense when there's an app-wide display for errors (like how GitHub has it under the nav bar) but when the error is localized to a component then I don't think it's necessary, since it's basically up to the component how the error should be handled (displayed somewhere, disable buttons, etc.).

As for utils, I get that by doing so it keeps the component file smaller, but in my opinion the purpose of a util is re-usability. I want to avoid passing in a bunch of arguments into a util function as that just makes the util very specific and hard to use. In this case, I think `isBalanceSufficient` is great as a util, but `getError` would be too specific to the component. `validateEditGas` could probably be moved to the container file, but we should figure out how to deal with the translation of error messages.

File: ui/app/actions.js (hunk 6 above)
Comment (alextsg) [reply ui/app/actions.js]: This is a weird case where the callback passed to `ethQuery.sendTransaction` gets called when the tx gets approved or rejected, and not when we'd expect (called after the tx is added to the list of unapproved tx'es). The loading spinner was being hidden unintentionally during approving/rejecting on the confirm step.

File: ui/app/components/tabs/tabs.component.js (hunk 9 above)
Comment (alextsg) [reply ui/app/components/tabs/tabs.component.js]: `React.Children` is a set of utils that can be used on `this.props.children` https://reactjs.org/docs/react-api.html#reactchildren

File: ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js (hunk 10 above)
Comment (alextsg) [reply ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js]: I was actually trying to avoid this so that the elements would be more HTML semantic:
```
<Tabs>
  <Tab />
  <Tab />
</Tabs>
```

File: ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js (hunk 11 above)
Comment (alextsg) [reply ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js]: Not here, because the `confirm-page-container-content__error-container` class adds padding around the `ConfirmPageContainerError` component. This makes the `ConfirmPageContainerError` more reusable and contained since the parent would be responsible for the spacing of the child. I might not want the same spacing/padding used elsewhere.

File: ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-summary/confirm-page-container-summary.component.js (hunk 12 above)
Comment (alextsg) [reply ui/app/components/confirm-page-container/confirm-page-container-content/confirm-page-container-summary/confirm-page-container-summary.component.js]: I did it this way because the default of the summary component is to show the subtitle, and hiding it is the exception. If I used `showSubtitle` as the boolean prop then I think it's a less convenient API since I'd have to add that in more places than not.

File: ui/app/components/pages/confirm-transaction-switch/confirm-transaction-switch.util.js (hunk 15 above)
Comment (alextsg) [reply ui/app/components/pages/confirm-transaction-switch/confirm-transaction-switch.util.js]: The tools I have for parsing data (abi-decoder and eth-method-registry) don't return anything from the deploy contract data, so I don't think there's a quick fix at our disposal. I asked kumavis about this as well and deploying a contract should be the only case when `txParams.to` is not defined.

File: ui/app/components/pages/confirm-transaction/confirm-transaction.component.js (hunk 17 above)
Comment (alextsg) [reply ui/app/components/pages/confirm-transaction/confirm-transaction.component.js]: I kept this method in the component file since it works with react-router props. I kinda see methods in the container file as having to do with modifying the data the component uses and not anything to do with the view, so I've been keeping react-router/view changes in the component and not the container.

File: ui/app/helpers/confirm-transaction/util.js (hunk 20 above)
Comment (alextsg) [reply ui/app/helpers/confirm-transaction/util.js]: I just found myself reusing these in various components and the .duck, and it got annoying having to move these functions/files around based on which component was using a function during debugging/refactoring (up one level when used by a parent, down one level when not used).

**Review by danjm — DISMISSED** [review]

**Review by danjm — APPROVED** [review]

**Comment by alextsg** [comment]
This closes #4298 closes #4635 closes #3414 closes #4404 closes #4481 closes #4429 closes #3824 closes #4430

**Comment by 1blockologist** [comment]
What is the process of getting this up into master/production? Didn't see a place to chat with the team

**Comment by bdresser** [comment]
@1blockologist this will bei n 4.9.0, should be out early next week! thanks for your patience.

**Comment by abitrolly** [comment]
Where are screenshots of before and after? Common, it is just Ctrl+Shift+PrtSc and Ctrl+V on GNOME. ;)
